platform/entanglement/src/services/LinkService.js

```diff
@@ -45,36 +45,24 @@ define(
             if (parentCandidate.getId() === object.getId()) {
                 return false;
             }
-            if (parentCandidate.getModel().composition.indexOf(object.getId()) !== -1) {
+            if ((parentCandidate.getModel().composition || []).indexOf(object.getId()) !== -1) {
                 return false;
             }
             return this.policyService.allow(
                 "composition",
                 parentCandidate.getCapability('type'),
                 object.getCapability('type')
-            );
+            ) && parentCandidate.hasCapability('composition');
         };
 
         LinkService.prototype.perform = function (object, parentObject) {
-            function findChild(children) {
-                var i;
-                for (i = 0; i < children.length; i += 1) {
-                    if (children[i].getId() === object.getId()) {
-                        return children[i];
-                    }
-                }
-            }
+            // Note that this was checked-for explicitly during validate step
```

Comment:
There's JSDoc for the interface; I believe I hoisted this out in the context of refactoring to use prototypes, #67. Since all three of MoveService/CopyService/LinkService expose the same interface this was otherwise redundant. But this doesn't really address the risk (a `validate` call might be missed for lots of reasons other than ignorance, and you're still left with the same this-is-hard-to-diagnose problem)

Added checks which throw exceptions if you try to perform a move/copy/link via a service which does not successfully validate. This is still fail-fast but more informative than some random exception with a comment, and has the benefit of matching validation logic consistently and not just erroring in cases where an exception happens to be hit.